// @frontegg/redux-store/node/auth/RolesState/saga.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rolesSagas = rolesSagas;
exports.rolesSagasMock = rolesSagasMock;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _effects = require("redux-saga/effects");
var _restApi = require("@frontegg/rest-api");
var _reducer = require("../reducer");
var _dummy = require("../dummy");
var _utils = require("../../utils");
const _excluded = ["callback"],
  _excluded2 = ["callback"],
  _excluded3 = ["callback"],
  _excluded4 = ["callback"],
  _excluded5 = ["callback"];
function* loadRolesAndPermissions({
  payload
}) {
  var _payload$silentLoadin;
  yield (0, _effects.put)(_reducer.actions.setRolesState({
    loading: !((_payload$silentLoadin = payload == null ? void 0 : payload.silentLoading) != null ? _payload$silentLoadin : false),
    error: null
  }));
  try {
    const result = yield (0, _effects.all)([(0, _effects.call)(_restApi.api.roles.getRoles), (0, _effects.call)(_restApi.api.roles.getPermissions), (0, _effects.call)(_restApi.api.roles.getPermissionCategories)]);
    const [roles, permissions, permissionCategories] = result;
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      roles,
      permissions,
      permissionCategories,
      loading: false
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      error: (0, _utils.errorHandler)(e),
      loading: false
    }));
  }
}
function* addRole(_ref) {
  let {
      payload: {
        callback
      }
    } = _ref,
    body = (0, _objectWithoutPropertiesLoose2.default)(_ref.payload, _excluded);
  yield (0, _effects.put)(_reducer.actions.setRolesState({
    saving: true
  }));
  try {
    const role = yield (0, _effects.call)(_restApi.api.roles.addRole, body);
    const roles = yield (0, _effects.call)(_restApi.api.roles.getRoles);
    yield (0, _effects.put)(_reducer.actions.loadRolesAndPermissions({
      silentLoading: true
    }));
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      roles,
      saving: false
    }));
    callback == null ? void 0 : callback(role);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      error: (0, _utils.errorHandler)(e),
      saving: false
    }));
    callback == null ? void 0 : callback(null, e);
  }
}
function* deleteRole(_ref2) {
  let {
      payload: {
        callback
      }
    } = _ref2,
    body = (0, _objectWithoutPropertiesLoose2.default)(_ref2.payload, _excluded2);
  yield (0, _effects.put)(_reducer.actions.setRolesState({
    saving: true
  }));
  try {
    yield (0, _effects.call)(_restApi.api.roles.deleteRole, body);
    yield (0, _effects.put)(_reducer.actions.loadRolesAndPermissions({
      silentLoading: true
    }));
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      saving: false
    }));
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      error: (0, _utils.errorHandler)(e),
      saving: false
    }));
    callback == null ? void 0 : callback(false, e);
  }
}
function* updateRole(_ref3) {
  let {
      payload: {
        callback
      }
    } = _ref3,
    body = (0, _objectWithoutPropertiesLoose2.default)(_ref3.payload, _excluded3);
  yield (0, _effects.put)(_reducer.actions.setRolesState({
    saving: true
  }));
  try {
    const role = yield (0, _effects.call)(_restApi.api.roles.updateRole, body);
    yield (0, _effects.put)(_reducer.actions.loadRolesAndPermissions({
      silentLoading: true
    }));
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      saving: false
    }));
    callback == null ? void 0 : callback(role);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      error: (0, _utils.errorHandler)(e),
      saving: false
    }));
    callback == null ? void 0 : callback(null, e);
  }
}
function* attachPermissionsToRole(_ref4) {
  let {
      payload: {
        callback
      }
    } = _ref4,
    body = (0, _objectWithoutPropertiesLoose2.default)(_ref4.payload, _excluded4);
  yield (0, _effects.put)(_reducer.actions.setRolesState({
    saving: true
  }));
  try {
    const role = yield (0, _effects.call)(_restApi.api.roles.attachPermissionsToRole, body);
    yield (0, _effects.put)(_reducer.actions.loadRolesAndPermissions({
      silentLoading: true
    }));
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      saving: false
    }));
    callback == null ? void 0 : callback(role);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      error: (0, _utils.errorHandler)(e),
      saving: false
    }));
    callback == null ? void 0 : callback(null, e);
  }
}
function* attachPermissionToRoles(_ref5) {
  let {
      payload: {
        callback
      }
    } = _ref5,
    body = (0, _objectWithoutPropertiesLoose2.default)(_ref5.payload, _excluded5);
  yield (0, _effects.put)(_reducer.actions.setRolesState({
    saving: true
  }));
  try {
    const permission = yield (0, _effects.call)(_restApi.api.roles.attachPermissionToRoles, body);
    yield (0, _effects.put)(_reducer.actions.loadRolesAndPermissions({
      silentLoading: true
    }));
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      saving: false
    }));
    callback == null ? void 0 : callback(permission);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      error: (0, _utils.errorHandler)(e),
      saving: false
    }));
    callback == null ? void 0 : callback(null, e);
  }
}
function* rolesSagas() {
  yield (0, _effects.takeLeading)(_reducer.actions.loadRolesAndPermissions, loadRolesAndPermissions);
  yield (0, _effects.takeEvery)(_reducer.actions.addRole, addRole);
  yield (0, _effects.takeEvery)(_reducer.actions.deleteRole, deleteRole);
  yield (0, _effects.takeEvery)(_reducer.actions.updateRole, updateRole);
  yield (0, _effects.takeEvery)(_reducer.actions.attachPermissionsToRole, attachPermissionsToRole);
  yield (0, _effects.takeEvery)(_reducer.actions.attachPermissionToRoles, attachPermissionToRoles);
}

/*********************************
 *  Preview Sagas
 *********************************/

function* loadRolesAndPermissionsMock({
  payload
}) {
  var _payload$silentLoadin2;
  yield (0, _effects.put)(_reducer.actions.setRolesState({
    loading: !((_payload$silentLoadin2 = payload == null ? void 0 : payload.silentLoading) != null ? _payload$silentLoadin2 : false),
    error: null
  }));
  try {
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      roles: _dummy.rolesAdminViewerDemo,
      loading: false
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setRolesState({
      error: (0, _utils.errorHandler)(e),
      loading: false
    }));
  }
}
function* rolesSagasMock() {
  yield (0, _effects.takeLeading)(_reducer.actions.loadRolesAndPermissions, loadRolesAndPermissionsMock);
}
